### DISCUSSION TITLE
Getting Issues in refund the donation

### DISCUSSION THREAD

**Discussion by dightinfotechcom** [impress-org/givewp#7426] [Developer Q&A]
Hello there

Please check this video first. 

https://www.loom.com/share/d9fa2cf7a062478d80cfe21b26099eab

While refund, I have added amount more that I donate. it is showing error. But it changed the status and amount for donation in Givewp admin dashboard. It should remain same and shows error that this refund is not possible or something. 

I have developed a plugin for custom payment gateway  "easymerchant".

everything working fine but while I am refunding it shows the above issues. I mean it is talking amount more than donation, status changed to refund also. 

can you explain why this is happening? also In my case of easymerchant, after clicking for refund it redirect to othe page while it should stay on same page and donation update. 

please look into this matter.

**Comment by kjohnson** [comment]
Hey @dightinfotechcom. Can you provide a link to the custom code for me to take a look at? If it's not too complicated I may be able to identify the issue. Do you have a link to a public repo or could you add me to a repo?

**Reply by dightinfotechcom** [reply]
Hi @kjohnson 

here is my custom code for refund donation 

public function refundDonation(Donation $donation): PaymentRefunded
    {
        $apiKey                 = lyfepay_give_get_api_key();
        $apiSecretKey           = lyfepay_give_get_api_secret_key();
        if (give_is_test_mode()) {
            // GiveWP is in test mode
            $apiUrl = 'https://stage-api.stage-easymerchant.io/api/v1';
        } else {
            $apiUrl = 'https://api.easymerchant.io/api/v1';
        }

        try {
            $checkStatusApi = wp_remote_post($apiUrl . '/charges/' . $donation->gatewayTransactionId, array(
                'method'    => 'GET',
                'headers'   => array(
                    'X-Api-Key'     => $apiKey,
                    'X-Api-Secret'  => $apiSecretKey,
                    'Content-Type'  => 'application/json',
                )
            ));
            $checkPaidUnsetteled = wp_remote_retrieve_body($checkStatusApi);
            $checkStatus = json_decode($checkPaidUnsetteled, true);
            $refundAmount = $checkStatus['data']['amount'] - $donation->amount->formatToDecimal();
            $body = json_encode([
                "charge_id" => $donation->gatewayTransactionId,
                'amount'    => $refundAmount,
            ]);

            if ($checkStatus['data']['status'] === 'Paid') {
                $response = wp_remote_post($apiUrl . '/refunds/', array(
                    'method'    => 'POST',
                    'headers'   => array(
                        'X-Api-Key'      => $apiKey,
                        'X-Api-Secret'   => $apiSecretKey,
                        'Content-Type'   => 'application/json',
                    ),
                    'body'               => $body,
                ));
                $response_body = wp_remote_retrieve_body($response);
                $response_data = json_decode($response_body, true);

                $donation->status = DonationStatus::REFUNDED();
                $donation->save();
                DonationNote::create([
                    'donationId' => $donation->id,
                    'content' => sprintf(esc_html__('Refund processed successfully. Reason: %s', 'lyfepay-give'), 'refunded by user')
                ]);
            }
            print_r($response_data['message']);

            echo "<script>
                setTimeout(() => {
                    window.history.go(-1);
                }, 1000);
                </script>";
            die();
            return new PaymentRefunded();
        } catch (\Exception $exception) {
            throw new PaymentGatewayException('Unable to refund. ' . $exception->getMessage(), $exception->getCode(), $exception);
        }
    }
    
    when I refund the donation It redirect to other page so I used this code to show message for successful refund. 
    then refund back to the page through js code
    
     print_r($response_data['message']);

            echo "<script>
                setTimeout(() => {
                    window.history.go(-1);
                }, 1000);
                </script>";
            die();
            
            By the way if you want to look the repo then here is git repo. 
            https://github.com/dightinfotechcom/EM-GiveWP

**Reply by dightinfotechcom** [reply]
Also please look into the matter how Can I refund more than the actual payment and it allows. this should not happen.
